DirectProgramming/DPC++FPGA/ReferenceDesigns/anr/src/qfp.hpp

```diff
@@ -0,0 +1,201 @@
+struct QFP {
+  QFP(const QFP&) = delete;
+  QFP& operator=(const QFP&) = delete;
+
+  // determine if the QFP can fit into a unsigned char or unsigned short (if
+  // not, default to an unsigned int)
+  static constexpr bool fits_in_uchar =
+    qfp_total_bits <= sizeof(unsigned char)*8;
+  static constexpr bool fits_in_ushort =
+    qfp_total_bits <= sizeof(unsigned short)*8;
+  
+  using qfp_type =
+    std::conditional_t<fits_in_uchar, unsigned char,
+      std::conditional_t<fits_in_ushort, unsigned short, unsigned>>;
+
+  // 32-bit floating point bits based on
+  // https://en.wikipedia.org/wiki/Single-precision_floating-point_format
+  static constexpr unsigned fp32_sign_bits = 1;
+  static constexpr unsigned fp32_exponent_bits = 8;
+  static constexpr unsigned fp32_mantissa_bits = 23;
+  static constexpr unsigned fp32_total_bits =
+      fp32_sign_bits + fp32_exponent_bits + fp32_mantissa_bits;
+  static constexpr int fp32_exponent_offset =
+    (1 << (fp32_exponent_bits-1)) - 1;
+  static constexpr unsigned fp32_exponent_mask = (1 << fp32_exponent_bits) - 1;
+  static constexpr unsigned fp32_mantissa_mask = (1 << fp32_mantissa_bits) - 1;
+
+  // A union for accesing the mantissa, exponent, and sign bits
+  typedef union {
+    float f;
+    struct {
+      unsigned mantissa   : fp32_mantissa_bits;
+      unsigned exponent   : fp32_exponent_bits;
+      unsigned sign       : fp32_sign_bits;
+    } parts;
+  } FloatCast;
+
+  // the number of mantissa bits for the QFP
+  static constexpr unsigned qfp_mantissa_bits =
+      qfp_total_bits - qfp_exponent_bits - is_signed;
+
+  // masks for the exponent and mantissa of the QFP
+  static constexpr unsigned qfp_mask = (1 << qfp_total_bits) - 1;
+  static constexpr unsigned qfp_exponent_mask = (1 << qfp_exponent_bits) - 1;
+  static constexpr unsigned qfp_mantissa_mask = (1 << qfp_mantissa_bits) - 1;
+  static constexpr int qfp_exponent_offset =
+      (1 << (qfp_exponent_bits - 1)) - 1;
+
+  // the difference in bits between the QFP and the 32-bit float
+  static constexpr unsigned mantissa_bit_diff =
+      fp32_mantissa_bits - qfp_mantissa_bits;
+
+  // static asserts
+  static_assert(fp32_total_bits == (sizeof(float) * 8));
+  static_assert(qfp_mantissa_bits <= fp32_mantissa_bits);
+  static_assert(qfp_exponent_bits > 0);
+  static_assert(qfp_mantissa_bits > 0);
+  static_assert(qfp_total_bits > qfp_exponent_bits);
+  
+  //
+  // convert from a 32-bit float to a QFP
+  //
+  static qfp_type FromFP32(float f) {
+    // use the float cast to get the parts of the FP32
+    FloatCast f_casted = {.f = f};
+    int fp32_sign = f_casted.parts.sign;
+    ac_int<fp32_exponent_bits + 1, true> fp32_exponent =
+        f_casted.parts.exponent;
+    ac_int<fp32_mantissa_bits, false> fp32_mantissa = f_casted.parts.mantissa;
+
+    // get the most significant qfp_mantissa_bits from the float's mantissa
+    auto qfp_mantissa =
+        (fp32_mantissa >> mantissa_bit_diff) & qfp_mantissa_mask;

```

Comment:
Should we be rounding the mantissa instead of truncating?  If you've decided it's not worth the extra area to do proper rounding, it's probably worth a comment to that effect here.